File: Tutorials/Codes/Modern_Control_Course/Chapter27/Lesson3/Chapter27_Lesson3.py
# Chapter27_Lesson3.py
# Internal Model Principle for step tracking in a continuous-time state-space plant.
# Required packages: numpy, scipy

import numpy as np
from scipy.integrate import solve_ivp
from scipy.signal import place_poles


def build_augmented_step_servo():
    """Plant:
        x_dot = A x + B u
        y     = C x
       Internal model for a constant reference:
        eta_dot = r - y
       Control:
        u = -Kx x + Ki eta
    """
    A = np.array([[0.0, 1.0],
                  [-2.0, -0.4]])
    B = np.array([[0.0],
                  [1.0]])
    C = np.array([[1.0, 0.0]])

    # Augmented state xa = [x1, x2, eta]^T, eta_dot = r - Cx.
    A_aug = np.block([
        [A, np.zeros((2, 1))],
        [-C, np.zeros((1, 1))]
    ])
    B_aug = np.vstack([B, [[0.0]]])

    desired_poles = np.array([-2.0 + 2.0j, -2.0 - 2.0j, -5.0])
    placed = place_poles(A_aug, B_aug, desired_poles)
    K_aug = placed.gain_matrix

    # Since u = -K_aug * [x; eta], and the integrator state is eta_dot = -Cx + r,
    # the third gain is negative for the convention u = -Kx + Ki eta.
    Kx = K_aug[:, :2]
    Ki = -K_aug[:, 2:3]
    return A, B, C, Kx, Ki, A_aug, B_aug


def simulate_step_response(t_final=8.0, reference=1.0):
    A, B, C, Kx, Ki, _, _ = build_augmented_step_servo()

    def closed_loop_rhs(t, xa):
        x = xa[:2].reshape(2, 1)
        eta = xa[2]
        y = float(C @ x)
        u = float(-Kx @ x + Ki * eta)
        xdot = A @ x + B * u
        etadot = reference - y
        return np.array([xdot[0, 0], xdot[1, 0], etadot])

    sol = solve_ivp(
        closed_loop_rhs,
        t_span=(0.0, t_final),
        y0=np.array([0.0, 0.0, 0.0]),
        max_step=0.01,
        rtol=1e-8,
        atol=1e-10,
    )

    x1 = sol.y[0, :]
    x2 = sol.y[1, :]
    eta = sol.y[2, :]
    y = x1
    u = -Kx[0, 0] * x1 - Kx[0, 1] * x2 + Ki[0, 0] * eta
    e = reference - y
    return sol.t, y, u, e, Kx, Ki


def solve_regulator_equations_for_constant_reference():
    """Solve the steady-state regulator equations for S = 0 and y_ref = w.
       A*pi + B*gamma = 0
       C*pi = 1
       This gives the steady-state x = pi*w and u = gamma*w.
    """
    A = np.array([[0.0, 1.0],
                  [-2.0, -0.4]])
    B = np.array([[0.0],
                  [1.0]])
    C = np.array([[1.0, 0.0]])

    M = np.block([
        [A, B],
        [C, np.zeros((1, 1))]
    ])
    rhs = np.array([[0.0], [0.0], [1.0]])
    sol = np.linalg.solve(M, rhs)
    pi = sol[:2, :]
    gamma = sol[2:, :]
    return pi, gamma


if __name__ == "__main__":
    t, y, u, e, Kx, Ki = simulate_step_response()
    pi, gamma = solve_regulator_equations_for_constant_reference()

    print("Kx =", Kx)
    print("Ki =", Ki)
    print("Regulator equation Pi =", pi.ravel())
    print("Regulator equation Gamma =", gamma.ravel())
    print("Final output y(T) =", y[-1])
    print("Final tracking error e(T) =", e[-1])
    print("Peak absolute control =", np.max(np.abs(u)))

File: Tutorials/Codes/Modern_Control_Course/Chapter27/Lesson3/Chapter27_Lesson3.m
% Chapter27_Lesson3.m
% Internal Model Principle in a linear state-space setting.
% Requires Control System Toolbox for place().
% Optional: creates a simple Simulink model if Simulink is available.

clear; clc; close all;

A = [0 1; -2 -0.4];
B = [0; 1];
C = [1 0];
D = 0;

% Step-reference internal model: eta_dot = r - y.
Aaug = [A zeros(2,1); -C 0];
Baug = [B; 0];

desired_poles = [-2+2i, -2-2i, -5];
Kaug = place(Aaug, Baug, desired_poles);

Kx = Kaug(1:2);
Ki = -Kaug(3);

fprintf('Kx = [%g  %g]\n', Kx(1), Kx(2));
fprintf('Ki = %g\n', Ki);

r = 1.0;
f = @(t,xa) [
    xa(2);
    -2*xa(1) - 0.4*xa(2) + (-Kx*[xa(1); xa(2)] + Ki*xa(3));
    r - C*[xa(1); xa(2)]
];

[t, xa] = ode45(f, [0 8], [0; 0; 0]);
y = xa(:,1);
u = -Kx(1)*xa(:,1) - Kx(2)*xa(:,2) + Ki*xa(:,3);
e = r - y;

figure;
plot(t, y, 'LineWidth', 1.5); grid on;
xlabel('Time (s)');
ylabel('Output y(t)');
title('Step Tracking with an Internal Model');

figure;
plot(t, e, 'LineWidth', 1.5); grid on;
xlabel('Time (s)');
ylabel('Tracking error e(t)');
title('Error Convergence');

fprintf('Final y(T) = %.8f\n', y(end));
fprintf('Final e(T) = %.8e\n', e(end));
fprintf('Peak |u| = %.8f\n', max(abs(u)));

% Regulator equations for constant reference:
% A*Pi + B*Gamma = 0, C*Pi = 1.
M = [A B; C 0];
rhs = [0; 0; 1];
sol = M \ rhs;
Pi = sol(1:2);
Gamma = sol(3);
fprintf('Pi = [%g  %g]^T\n', Pi(1), Pi(2));
fprintf('Gamma = %g\n', Gamma);

% Optional Simulink skeleton: integrator state eta and State-Space plant block.
if license('test','SIMULINK')
    model = 'Chapter27_Lesson3_Simulink';
    if bdIsLoaded(model)
        close_system(model, 0);
    end
    new_system(model);
    open_system(model);

    add_block('simulink/Sources/Step', [model '/Reference'], ...
        'Position', [40 80 80 110], 'Time', '0', 'Before', '0', 'After', '1');
    add_block('simulink/Continuous/State-Space', [model '/Plant'], ...
        'Position', [320 70 430 130], ...
        'A', mat2str(A), 'B', mat2str(B), 'C', mat2str(C), 'D', mat2str(D));
    add_block('simulink/Math Operations/Sum', [model '/Error r-y'], ...
        'Position', [130 75 160 115], 'Inputs', '+-');
    add_block('simulink/Continuous/Integrator', [model '/Internal Model Integrator'], ...
        'Position', [190 75 230 115]);
    add_block('simulink/Sinks/Scope', [model '/Output Scope'], ...
        'Position', [500 75 540 115]);

    % Full feedback interconnection usually needs gain and sum blocks for
    % u = -Kx + Ki*eta. This skeleton shows the internal-model signal path.
    add_line(model, 'Reference/1', 'Error r-y/1');
    add_line(model, 'Error r-y/1', 'Internal Model Integrator/1');
    add_line(model, 'Plant/1', 'Output Scope/1');

    save_system(model);
    fprintf('Created Simulink skeleton: %s.slx\n', model);
end
